enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+### Goals
+
+- Prepare OpenShift components to accommodate the CCMs instead of deprecated in-tree cloud providers.
+- Provide the means to allow management of generic `CCM` component based on the platform specific upstream implementation.
+- Define and implement an upgrading and downgrading paths between the in-tree and the out-of-tree cloud provider configurations.
+- Ensure the transition will ensure the feature parity between the in-tree and the out-of-tree provider.
+- Consider support for multiple providers within a single cluster as a possible future configuration (and make sure there are no clashes between provider configurations).
+
+### Non-Goals
+
+- Force an immediate transition for all cloud providers from in-tree support to their out-of-tree counterparts.
+- [CSI driver migration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) is out of scope of this work.
+
+## Proposal
+
+Our main goal is to start using Cloud Controller Manager in OpenShift 4, and make a seamless transition for all currently supported providers that require `cloud provider` interface: OpenStack, AWS, Azure, GCP and vSphere.
+
+To maintain the lifecycle of the CCM we want to implement a cluster operator called `cluster-cloud-controller-manager-operator`, that will handle all administrative tasks: deploy, restore, upgrade, and so on.
+
+### User Stories
+
+#### Story 1
+
+As a cloud developer, I’d like to improve support for cloud features for OpenShift. I'd like to develop and test new cloud providers which do not ship with default Kubernetes distribution, and require support for external cloud controller manager.
+
+#### Story 2
+
+As a developer I'd like to implement, build and release fixes for cloud provider controllers independently from the Kubernetes core, and assume they will first land upstream and then will be carried over into OpenShift distribution with less effort.
+
+#### Story 3
+
+As an OpenShift developer responsible for the cloud controllers, I want to share more code with upstream Kubernetes in order to ease feature and bug fix contributions both ways.
+
+#### Story 4
+
+We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.
+
+### Implementation Details
+
+The `cluster-cloud-controller-manager-operator` implementation will be hosted within its own OpenShift [repository](https://github.com/openshift/cluster-cloud-controller-manager-operator). The operator will manage `CCM` provisioning for supported cloud providers.
+
+- The Operator and all platform cloud controllers will be provisioned in `openshift-cloud-controller-manager` namespace.
```

Comment:
Move this sentence to where you're describing the ccm-o details.

Reply:
This is also a strange factoring. Usually the operands (= cloud controllers) live in a different namespace than the operator.

Reply:
This is my fault, it was originally two namespaces but I was asking why we need two, MAO all lives in one namespace and that seems to work, so I had asked what the reason for two namespaces would be.

If there's a convention for having operators and operands in separate namespaces we can do that here, is there a particular reason for it other than convention?

Reply:
We would also avoid questions about why an empty namespace exists in every cluster, while it does not host any operands in it. Having a single namespace seems to be clean an consistent to me too.

Reply:
empty namespace doesn't cost anything